[Thia Su Mian] iP

tsumian opened the pull request:
# Duke
> "With Duke you will not need another pay-to-use To Do list." - Thia Su Mian

Duke can do:
- To do 
- Deadlines
- Events
- Find tasks
- Mark and Unmark tasks

**SO SIMPLE!!** :sunglasses::sunglasses:
For more emojis click [here](https://github.com/ikatyang/emoji-cheat-sheet/blob/master/README.md)

Let me show you how simple the `main` code is
# 
```java
    public static void main(String[] args) {
        new Duke("src/main/java/data/savedTasks.txt").dukeRunner();
    }
```

Have I:
- [ ] Convinced
- [ ] You
- [ ] Yet?

If not:
1. Let me know what I can do.
2. Leave.

Commit:
Add Gradle support

tsumian committed:
Level-1

tsumian committed:
Level-2

tsumian committed:
Level-2

tsumian committed:
Level-3

tsumian committed:
Todo, Events and Deadlines

tsumian committed:
Level-4

tsumian committed:
A-TextUiTesing

tsumian committed:
Level-5

tsumian committed:
Level-6

tsumian committed:
no message

tsumian committed:
Revert "no message"

This reverts commit a4c25003620011cac036b02a71825ff61f3cb27c.

tsumian committed:
Add save to disk function

tsumian committed:
Add Times and Dates to Events/Deadlines

tsumian committed:
Add packages Duke and Exception

tsumian committed:
Add unit tests for task and ListStorage

tsumian committed:
Add JAR file

tsumian committed:
Update JavaDoc

tsumian committed:
Update JavaDoc

tsumian committed:
Update Coding Standard and JavaDoc

tsumian committed:
Update Changes

tsumian committed:
Merge commit 'b459b9e8cfc49122372781708980c992b88440d3'

tsumian committed:
Add find function

tsumian committed:
Merge branch 'add-gradle-support'

petermonky commented on src/main/java/Duke/Commands.java:
```diff
@@ -0,0 +1,168 @@
+     * Constructor for Commands. Commands contain a ListStorage
+     * and Printer class.
+     *
+     * @param listStorage Specifies a ListStorage to store commands
```
Shouldn't there be a period (.) at the end of this `@param` description?

petermonky commented on src/main/java/Duke/Commands.java:
```diff
@@ -0,0 +1,168 @@
+import Duke.Exceptions.EmptyMessageException;
+import Duke.Exceptions.WrongDateFormatException;
+
+public class Commands {
```
I like how you implemented your commands as separate methods of the `Commands` class! Though I am a little wary about how this might affect scalability.

petermonky commented on src/main/java/Duke/Commands.java:
```diff
@@ -0,0 +1,168 @@
+            throw new WrongDateFormatException("Wrong time format");
+        } else if (!deadline[1].contains("AM") && !deadline[1].contains("PM")) {
+            throw new WrongDateFormatException("Wrong time format");
+        }
```
It's cool to see how you handled exceptions separately for date and time! I can definitely see how it would help the user.

petermonky commented on src/main/java/Duke/Disk.java:
```diff
@@ -0,0 +1,87 @@
+import java.io.IOException;
+import java.util.Scanner;
+
+public class Disk {
```
Could this class have a JavaDoc that details what the class is for? At surface level, it's quite hard to imagine what it would be used for.

petermonky commented on src/main/java/Duke/Duke.java:
```diff
@@ -0,0 +1,117 @@
+                break;
+            } catch (WrongCommandException e) {
+                myPrinter.printExceptions(e);
+            }
```
If I'm not wrong, this is to ensure that Duke continues to run even after accepting a wrong command? I like the logic of this portion.

petermonky commented on src/main/java/Duke/Duke.java:
```diff
@@ -0,0 +1,117 @@
+                    } else {
+                        throw new WrongCommandException("Invalid Command");
+                    }
+                }
```
Are commands intepreted with `contains()`? How would this affect the programme if the user types a command with multiple keywords? e.g. `event todo`?

petermonky commented on src/main/java/Duke/Exceptions/DukeException.java:
```diff
@@ -0,0 +1,7 @@
+package Duke.Exceptions;
+
+public class DukeException extends Exception{
```
Shouldn't there be a space between `Exception` and `{`?

petermonky commented on src/main/java/Duke/ListStorage.java:
```diff
@@ -0,0 +1,89 @@
+public class ListStorage {
+    /** Single List to store Tasks */
+    public List<Task> myTasks;
+    String SPACING = "    ";
```
I think this constant can be declared as `static final`, considering its use case.

petermonky commented on src/main/java/Duke/ListStorage.java:
```diff
@@ -0,0 +1,89 @@
+     * @return description of task.
+     */
+    public String printTask(int taskNumber) {
+        Task currentTask = myTasks.get(taskNumber - 1);
```
Could `myTasks.get(taskNumber - 1)` be replaced with `findTask(taskNumber)` as declared below?

petermonky commented on src/main/java/Duke/Printer.java:
```diff
@@ -0,0 +1,124 @@
+        System.out.println(BORDER + SPACING +
+                "Hello! I'm Duke\n" + SPACING +
+                "What can I do for you?\n" +
+                BORDER);
```
Would it be better to add a line break before the `+` operator to improve readability? i.e. from

```java
System.out.println(BORDER + SPACING +
        "Hello! I'm Duke\n" + SPACING +
        "What can I do for you?\n" +
        BORDER);
```

to

```java
System.out.println(BORDER + SPACING
        + "Hello! I'm Duke\n" + SPACING
        + "What can I do for you?\n"
        + BORDER);
```

petermonky commented on src/main/java/Duke/Printer.java:
```diff
@@ -0,0 +1,124 @@
+     * @param task Task added to storage.
+     * @param numberOfTasks Number of Tasks in storage.
+     */
+    public void  printTask(Task task, int numberOfTasks) {
```
Should the double space between `void` and `printTask` be a single space?

petermonky commented on src/main/java/Duke/Printer.java:
```diff
@@ -0,0 +1,124 @@
+                + SPACING
+                + "Now you have " + numberOfTasks + " tasks in the list.\n"
+                + BORDER);
+    }
```
I like how you how you formatted this portion, I can readily visualise what Duke would actually print.

petermonky commented on src/main/java/Duke/Task.java:
```diff
@@ -0,0 +1,58 @@
+     *
+     * @return 1 if it is done, 0 if not done.
+     */
+    public char isDone() {
```
I think the name of this method might be a bit misleading; might lead one to believe that it returns a boolean. Perhaps the name could be changed to more clearly reflect this?

petermonky reviewed (commented):
Overall, a very nicely structured project; some clever logic here and there. Other than some minor coding standard deviations, a solid piece of work. Consider splitting the commands up to individual subclasses for scalability.

tsumian committed:
Add gradle support to automate Unit tests

tsumian committed:
Add CheckStyle and made relevant changes

tsumian committed:
Add gui foundation

tsumian committed:
Add Duke response to GUI

tsumian committed:
Add assertions to commands methods

tsumian committed:
Improve code quality of Duke parseCommand method

Instead of using contains(), where users could input something like
"tododeadline <task>" and Duke still allows it, using switch statements
will only process the first argument as the user-intended command.

tsumian committed:
Merge pull request #2 from tsumian/branch-A-CodeQuality

Improve code quality of Duke parseCommand method

tsumian committed:
Resolv conflicts

tsumian committed:
Merge pull request #1 from tsumian/branch-A-Assertions

Add assertions to commands methods

tsumian committed:
Set up CI

tsumian committed:
Merge github.com:tsumian/ip

tsumian committed:
Add CI

tsumian committed:
Add CI

tsumian committed:
Add statistic functionality

Duke does not support statistics function such as querying completed task in a certain timeframe.

Adding statistics function allows users to track completed tasks within an enquiry period.

cmdStats method added to handle keyword 'stats'. It parses the command and enquiry date, allowing
Statistics class to process and compare completed tasks with the date in question. Any tasks done after
the enquiry date up till current point of time will be displayed in the UI as a list.

Using a separate method to handle different command is always desirable and neat. Keeps methods
short. Using a new Statistics class abstracts away the low level details of implementation such as
comparing using LocalDate and the values compareTo() returns, to avoid potential confusion.

tsumian committed:
Merge github.com:tsumian/ip

tsumian committed:
Set theme jekyll-theme-hacker

tsumian committed:
Update README.md

Current README is a skeleton UG

Updated README to act as a more detailed UG. UG includes features and brief usage of task bot.

tsumian committed:
Add OS independent specified file path

Current bot loads and saves in a hardcoded file path.

Changed file path to relative file path and OS independent file path so that errors will not occur
when jar file is run.

tsumian committed:
Add style

tsumian committed:
Edit Javafx files

tsumian committed:
Add jar file

Initial jar file had errors preventing GUI from opening when run in command line

Edited the resource files such that fxmlLoader could load properly.

tsumian committed:
Add better GUI with some color schemes

tsumian committed:
Add Ui Screenshot to docs

tsumian committed:
Merge github.com:tsumian/ip
